File: generated_code/claude/12_secrets_aws.py
import boto3
from botocore.exceptions import ClientError, NoCredentialsError
import os
from pathlib import Path

class S3Uploader:
    """
    A class to handle file uploads to AWS S3 bucket.
    """
    
    def __init__(self, bucket_name, region_name='us-east-1'):
        """
        Initialize S3 client.
        
        Args:
            bucket_name: Name of the S3 bucket
            region_name: AWS region (default: us-east-1)
        """
        self.bucket_name = bucket_name
        self.s3_client = boto3.client('s3', region_name=region_name)
    
    def upload_file(self, file_path, object_name=None):
        """
        Upload a single file to S3 bucket.
        
        Args:
            file_path: Path to the file to upload
            object_name: S3 object name. If not specified, file_path basename is used
            
        Returns:
            True if file was uploaded, else False
        """
        if object_name is None:
            object_name = os.path.basename(file_path)
        
        try:
            self.s3_client.upload_file(file_path, self.bucket_name, object_name)
            print(f"Successfully uploaded {file_path} to {self.bucket_name}/{object_name}")
            return True
        except FileNotFoundError:
            print(f"Error: The file {file_path} was not found")
            return False
        except NoCredentialsError:
            print("Error: AWS credentials not found")
            return False
        except ClientError as e:
            print(f"Error uploading file: {e}")
            return False
    
    def upload_file_with_metadata(self, file_path, object_name=None, metadata=None):
        """
        Upload a file with custom metadata.
        
        Args:
            file_path: Path to the file
            object_name: S3 object name
            metadata: Dict of metadata key-value pairs
        """
        if object_name is None:
            object_name = os.path.basename(file_path)
        
        extra_args = {}
        if metadata:
            extra_args['Metadata'] = metadata
        
        try:
            self.s3_client.upload_file(
                file_path, 
                self.bucket_name, 
                object_name,
                ExtraArgs=extra_args
            )
            print(f"Successfully uploaded {file_path} with metadata")
            return True
        except ClientError as e:
            print(f"Error: {e}")
            return False
    
    def upload_with_progress(self, file_path, object_name=None):
        """
        Upload file with progress callback.
        """
        if object_name is None:
            object_name = os.path.basename(file_path)
        
        file_size = os.path.getsize(file_path)
        
        def progress_callback(bytes_transferred):
            percentage = (bytes_transferred / file_size) * 100
            print(f"Upload progress: {percentage:.2f}%", end='\r')
        
        try:
            self.s3_client.upload_file(
                file_path,
                self.bucket_name,
                object_name,
                Callback=progress_callback
            )
            print(f"\nSuccessfully uploaded {file_path}")
            return True
        except ClientError as e:
            print(f"Error: {e}")
            return False
    
    def upload_directory(self, directory_path, s3_prefix=''):
        """
        Upload all files in a directory to S3.
        
        Args:
            directory_path: Path to the directory
            s3_prefix: Prefix to add to S3 object names
        """
        path = Path(directory_path)
        
        if not path.is_dir():
            print(f"Error: {directory_path} is not a directory")
            return False
        
        uploaded_count = 0
        for file_path in path.rglob('*'):
            if file_path.is_file():
                relative_path = file_path.relative_to(path)
                s3_key = f"{s3_prefix}/{relative_path}" if s3_prefix else str(relative_path)
                s3_key = s3_key.replace('\\', '/')  # Handle Windows paths
                
                if self.upload_file(str(file_path), s3_key):
                    uploaded_count += 1
        
        print(f"Uploaded {uploaded_count} files from {directory_path}")
        return True
    
    def upload_with_public_read(self, file_path, object_name=None):
        """
        Upload file with public read access.
        """
        if object_name is None:
            object_name = os.path.basename(file_path)
        
        try:
            self.s3_client.upload_file(
                file_path,
                self.bucket_name,
                object_name,
                ExtraArgs={'ACL': 'public-read'}
            )
            print(f"Uploaded {file_path} with public read access")
            return True
        except ClientError as e:
            print(f"Error: {e}")
            return False


# Usage Examples
if __name__ == "__main__":
    # Initialize uploader
    uploader = S3Uploader(bucket_name='your-bucket-name', region_name='us-east-1')
    
    # Example 1: Upload a single file
    uploader.upload_file('local-file.txt', 'remote-file.txt')
    
    # Example 2: Upload with metadata
    uploader.upload_file_with_metadata(
        'document.pdf',
        'documents/document.pdf',
        metadata={'author': 'John Doe', 'version': '1.0'}
    )
    
    # Example 3: Upload with progress tracking
    uploader.upload_with_progress('large-file.zip')
    
    # Example 4: Upload entire directory
    uploader.upload_directory('my-folder', s3_prefix='uploads/my-folder')
    
    # Example 5: Upload with public read access
    uploader.upload_with_public_read('public-image.jpg', 'images/public-image.jpg')


# Key Features:

# Simple file upload - Basic upload functionality
# Metadata support - Add custom metadata to files
# Progress tracking - Monitor upload progress for large files
# Directory uploads - Upload entire folders recursively
# Public access - Set files to be publicly readable
# Error handling - Handles common errors gracefully

# Prerequisites:

# Install boto3: pip install boto3
# Configure AWS credentials using one of these methods:

# AWS CLI: aws configure
# Environment variables: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY
# IAM role (if running on EC2/ECS)



# Quick Start:
# pythonuploader = S3Uploader('my-bucket-name')
# uploader.upload_file('myfile.txt')

File: generated_code/deepseek/12_secrets_aws.js
// 1. Using AWS SDK for JavaScript (Node.js)
// First, install the AWS SDK:

// bash
// npm install @aws-sdk/client-s3
// Basic Upload Code:
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
import fs from 'fs';

// Configure AWS credentials (prefer using environment variables or AWS config)
const s3Client = new S3Client({
  region: 'us-east-1',
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  }
});

async function uploadFileToS3(filePath, bucketName, keyName) {
  try {
    const fileContent = fs.readFileSync(filePath);
    
    const params = {
      Bucket: bucketName,
      Key: keyName, // File name in S3
      Body: fileContent,
      // Optional metadata
      Metadata: {
        'uploaded-by': 'nodejs-script'
      }
    };

    const command = new PutObjectCommand(params);
    const response = await s3Client.send(command);
    
    console.log(`File uploaded successfully. ETag: ${response.ETag}`);
    return response;
  } catch (error) {
    console.error('Error uploading file:', error);
    throw error;
  }
}

// Usage
uploadFileToS3('./myfile.txt', 'my-bucket', 'uploads/myfile.txt');
// Upload with Progress Tracking:

import { Upload } from '@aws-sdk/lib-storage';
import fs from 'fs';

async function uploadWithProgress(filePath, bucketName, keyName) {
  const fileStream = fs.createReadStream(filePath);
  
  const upload = new Upload({
    client: s3Client,
    params: {
      Bucket: bucketName,
      Key: keyName,
      Body: fileStream,
    }
  });

  upload.on('httpUploadProgress', (progress) => {
    console.log(`Uploaded: ${progress.loaded} / ${progress.total} bytes`);
  });

  const result = await upload.done();
  console.log('Upload complete:', result.Location);
  return result;
}
// 2. Using Python (boto3)
// First, install boto3:

// bash
// pip install boto3
// python
// import boto3
// from botocore.exceptions import NoCredentialsError

// def upload_file_to_s3(file_name, bucket, object_name=None):
//     """Upload a file to an S3 bucket"""
    
//     # If S3 object_name was not specified, use file_name
//     if object_name is None:
//         object_name = file_name
    
//     # Create S3 client
//     s3_client = boto3.client('s3')
    
//     try:
//         # Upload the file
//         response = s3_client.upload_file(
//             file_name, 
//             bucket, 
//             object_name,
//             ExtraArgs={
//                 'ACL': 'private',  # or 'public-read' for public files
//                 'ContentType': 'text/plain'  # Set appropriate content type
//             }
//         )
//         print(f"File {file_name} uploaded to {bucket}/{object_name}")
//         return True
//     except FileNotFoundError:
//         print(f"The file {file_name} was not found")
//         return False
//     except NoCredentialsError:
//         print("Credentials not available")
//         return False

// # Usage
// upload_file_to_s3('myfile.txt', 'my-bucket', 'uploads/myfile.txt')
// Upload with Metadata and Tags:
// python
// def upload_file_with_metadata(file_name, bucket, object_name):
//     s3_client = boto3.client('s3')
    
//     with open(file_name, 'rb') as file:
//         response = s3_client.put_object(
//             Bucket=bucket,
//             Key=object_name,
//             Body=file,
//             Metadata={
//                 'author': 'john-doe',
//                 'department': 'engineering'
//             },
//             Tagging='Environment=Production&Project=WebApp'
//         )
    
//     return response
// 3. Using AWS CLI
// First, install AWS CLI and configure credentials:

// bash
// aws configure
// Upload a single file:

// bash
// aws s3 cp myfile.txt s3://my-bucket/uploads/myfile.txt
// Upload entire directory:

// bash
// aws s3 sync ./local-folder s3://my-bucket/remote-folder
// 4. Using Presigned URLs (Client-side upload)
// Server-side: Generate presigned URL
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';

async function generatePresignedUrl(bucketName, keyName) {
  const command = new PutObjectCommand({
    Bucket: bucketName,
    Key: keyName,
    ContentType: 'application/octet-stream'
  });
  
  const url = await getSignedUrl(s3Client, command, { expiresIn: 3600 });
  return url;
}

// Client-side: Upload using fetch
async function uploadToPresignedUrl(file, presignedUrl) {
  const response = await fetch(presignedUrl, {
    method: 'PUT',
    body: file,
    headers: {
      'Content-Type': file.type
    }
  });
  
  if (response.ok) {
    console.log('Upload successful');
  } else {
    console.error('Upload failed');
  }
}
// # 5. Multipart Upload for Large Files
import { S3Client, CreateMultipartUploadCommand, 
         UploadPartCommand, CompleteMultipartUploadCommand } from '@aws-sdk/client-s3';

async function uploadLargeFile(filePath, bucketName, keyName) {
  const fileStream = fs.createReadStream(filePath);
  const fileSize = fs.statSync(filePath).size;
  const partSize = 5 * 1024 * 1024; // 5MB parts
  
  // Initiate multipart upload
  const createParams = { Bucket: bucketName, Key: keyName };
  const createCommand = new CreateMultipartUploadCommand(createParams);
  const { UploadId } = await s3Client.send(createCommand);
  
  let partNumber = 1;
  const parts = [];
  
  // Upload parts
  for (let start = 0; start < fileSize; start += partSize) {
    const end = Math.min(start + partSize, fileSize);
    const partBuffer = Buffer.alloc(end - start);
    fileStream.read(partBuffer);
    
    const uploadParams = {
      Bucket: bucketName,
      Key: keyName,
      UploadId,
      PartNumber: partNumber,
      Body: partBuffer
    };
    
    const uploadCommand = new UploadPartCommand(uploadParams);
    const { ETag } = await s3Client.send(uploadCommand);
    
    parts.push({ ETag, PartNumber: partNumber });
    partNumber++;
  }
  
  // Complete upload
  const completeParams = {
    Bucket: bucketName,
    Key: keyName,
    UploadId,
    MultipartUpload: { Parts: parts }
  };
  
  const completeCommand = new CompleteMultipartUploadCommand(completeParams);
  return await s3Client.send(completeCommand);
}
// Environment Variables Setup
// Create a .env file:

// bash
// AWS_ACCESS_KEY_ID=your_access_key
// AWS_SECRET_ACCESS_KEY=your_secret_key
// AWS_REGION=us-east-1
// AWS_BUCKET_NAME=your-bucket-name
// Security Best Practices:
// Never hardcode credentials - Use environment variables or IAM roles

// Use IAM roles for applications running on AWS

// Set appropriate permissions using bucket policies

// Enable encryption for sensitive data

// Validate file types and sizes server-side

// Use presigned URLs for client-side uploads

// Choose the method based on your use case:

// Node.js/Python SDKs: For server-side applications

// AWS CLI: For manual/uploads or scripting

// Presigned URLs: For direct client-to-S3 uploads

// Multipart Upload: For large files (>100MB)